### PR TITLE
#161779782 Build REST API End-point for Social Login

### PR DESCRIPTION
**Pull request by Daymorelah**
#### What does this PR do?
Build REST end-point to handle social login feature
#### Description of Task to be completed?
* Build login controller to handle a **GET** request to `auth/social_XXX` URL
* Write unit tests for the social login controller
* Write integration test for the social login controller
* Test route with postman to ensure controller is working as expected
#### How should this be manually tested?
* Fork the repo and `cd` into the project folder
* Run `npm install` to install all required packages
* Run 'npm test` to run tests on the app.
#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/161779782

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,7 +28,7 @@
     "errorhandler": "1.5.0",
     "express": "4.16.3",
     "express-jwt": "5.3.1",
-    "express-session": "1.15.6",
+    "express-session": "^1.15.6",
```
Comment (kingisaac95) [inline package.json]: Lock version please

File: package.json [hunk 2]
```diff
@@ -37,8 +37,11 @@
     "mongoose-unique-validator": "2.0.1",
     "morgan": "1.9.1",
     "nodemon": "1.12.6",
-    "passport": "0.4.0",
+    "passport": "^0.4.0",
```
Comment (kingisaac95) [inline package.json]: Ditto!

File: src/models/index.js [hunk 3]
```diff
@@ -4,8 +4,10 @@ import Sequelize from 'sequelize';
 import databaseConfig from '../config/config';
 
 const basename = path.basename(__filename);
-const env = process.env.NODE_ENV || 'development';
-const configEnv = databaseConfig[env];
+// const env = process.env.NODE_ENV || 'development';
```
Comment (kingisaac95) [inline src/models/index.js]: Remove comments

File: src/controller/socialLoginController.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  static async facebookLogin(req, res) {
+    const verifyWithObjectProperties = {
+      verifyWith: 'email',
+      profile: req.user,
+      socialMedia: 'facebook',
+      res,
+    };
+    try {
+      return SocialLogin.verifyAndLoginUser(verifyWithObjectProperties);
+    } catch (error) {
```
Comment (kingisaac95) [inline src/controller/socialLoginController.js:71]: Error response not used!

File: src/controller/socialLoginController.js [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  static async googleLogin(req, res) {
+    const verifyWithObjectProperties = {
+      verifyWith: 'email',
+      profile: req.user,
+      socialMedia: 'google',
+      res,
+    };
+    try {
+      return SocialLogin.verifyAndLoginUser(verifyWithObjectProperties);
+    } catch (error) {
```
Comment (kingisaac95) [inline src/controller/socialLoginController.js:96]: Ditto!

File: src/controller/socialLoginController.js (hunk 4 above)
Comment (Daymorelah) [reply src/controller/socialLoginController.js:71]: The error that will be caught will be an internal server error. So I am just sending an "Internal server error" error message to the front-end.

File: src/controller/socialLoginController.js (hunk 5 above)
Comment (Daymorelah) [reply src/controller/socialLoginController.js:96]: Ditto!

File: src/controller/socialLoginController.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+      if (error.errors) {
+        return helperMethods.sequelizeValidationError(res, error);
+      }
+      return helperMethods.serverError(res);
+    }
+  }
```
Comment (chinemelu) [inline src/controller/socialLoginController.js:127]: There are a lot of return functions from line 120 to line 126. You can reduce them to reduce code complexity.

File: src/migrations/02-create-users.js [hunk 7]
```diff
@@ -13,7 +13,7 @@ module.exports = {
     },
     email: {
       type: Sequelize.STRING,
-      unique: true,
+      unique: false,
```
Comment (chinemelu) [inline src/migrations/02-create-users.js]:
```suggestion
      unique: true,
```

File: test/unitTest/user.test.js [hunk 8]
```diff
@@ -1,9 +1,10 @@
 
 import chai from 'chai';
-import models from '../../src/models';
+import models from '../../src/models'
```
Comment (Oluwasegun-AA) [inline test/unitTest/user.test.js]: Good job @Daymorelah, but take to heart little details as this
``` suggestion
import models from '../../src/models';
```

File: src/migrations/20181202100524-create-users.js [hunk 9]
```diff
@@ -13,7 +13,7 @@ export default {
     },
     email: {
       type: Sequelize.STRING,
-      unique: true,
+      unique: false,
```
Comment (Oluwasegun-AA) [inline src/migrations/20181202100524-create-users.js]: Also, @Daymorelah, I feel each user should have a unique email address.. to eliminate situations where a user registers more than once using the same email.

```suggestion
      unique: true
```

File: src/controller/socialLoginController.js (hunk 6 above)
Comment (Daymorelah) [reply src/controller/socialLoginController.js:127]: All return statements are necessary for the code to work as expected.

File: src/migrations/02-create-users.js (hunk 7 above)
Comment (Daymorelah) [reply src/migrations/02-create-users.js]: Thank you. This has been fixed.

File: test/unitTest/user.test.js (hunk 8 above)
Comment (Daymorelah) [reply test/unitTest/user.test.js]: Thank you. This has been fixed.

File: src/controller/socialLoginController.js (hunk 5 above)
Comment (Daymorelah) [reply src/controller/socialLoginController.js:96]: This has been fixed. Thank you.

**Review by kingisaac95 — APPROVED** [review]

**Pull request merged**